### DISCUSSION TITLE
SharedResourceWarning when CRD has app.kubernetes.io/instance label

### DISCUSSION THREAD

**Discussion by EdwardCooke** [argoproj/argo-cd#8545] [Q&A] [answered]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
In version 2.3.0-rc5 when applying a custom resource definition with an `app.kubernetes.io/instance` label it shows an error stating it is part of 2 different applications, the argo application and the value assigned to that field.

**To Reproduce**

 Apply the following kustomization using Argo
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- https://download.elastic.co/downloads/eck/2.0.0/crds.yaml
- https://download.elastic.co/downloads/eck/2.0.0/operator.yaml

namespace: elastic-system
```

**Expected behavior**

No warning about being part of a second application

**Screenshots**
![warnings](https://user-images.githubusercontent.com/81339027/154001224-ca57aacb-dd2f-4d16-9853-a5679d22d6c6.png)

**Version**

```shell
{
    "Version": "v2.3.0-rc5+aa54aa7",
    "BuildDate": "2022-02-05T00:14:19Z",
    "GitCommit": "aa54aa7eda70f5e049c9e25df75221d899efff4b",
    "GitTreeState": "clean",
    "GoVersion": "go1.17.6",
    "Compiler": "gc",
    "Platform": "linux/arm64",
    "KsonnetVersion": "vdev-2022-02-05T00:05:16+0000",
    "KustomizeVersion": "v4.4.1 2021-11-11T23:36:27Z",
    "HelmVersion": "v3.7.2+g663a896",
    "KubectlVersion": "v0.23.1",
    "JsonnetVersion": "v0.18.0"
}
```

**Logs**

**Comment by jessesuen** [comment, marked as answer]
We inject the app name into `app.kubernetes.io/instance`. Sometimes helm charts conflict with the `app.kubernetes.io/instance` label which is what is happening here. The crd.yaml is hardwiring a `app.kubernetes.io/instance` label:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: null
  labels:
    app.kubernetes.io/instance: 'elastic-operator'
```

In which case you should consider using annotation based tracking (which will become the default in the future) so that there is no conflict.

https://argo-cd.readthedocs.io/en/latest/user-guide/resource_tracking/#additional-tracking-methods-via-an-annotation

**Reply by roy-work** [reply]
It seems like if a Helm chart renders an `app.kubernetes.io/instance` label, ArgoCD will *usually* overwrite it with what it uses for tracking. That's fine.

But it seems like that overwriting doesn't happen for CRDs … why?

The Helm chart I have does `app.kubernetes.io/instance: {{ .Release.Name }}`¹, but consistently/for multiple resources. ArgoCD seems to overwrite the output, except on the CRD, and so the CRD then gets the `SharedResourceWarning`. But why is the CRD different here?

Edit: I'm even more confused; I followed @HariSekhon's lead & removed the label … but ArgoCD doesn't seem to add anything to the resource? How does Argo know its the one tracking the CRD…?

¹(which Helm [seems to recommend](https://helm.sh/docs/chart_best_practices/labels/#standard-labels))

**Comment by HariSekhon** [comment]
I've worked around this issue by dynamically removing the label from the helm chart in Kustomize using a json patch, eg:

https://github.com/HariSekhon/Kubernetes-configs/blob/master/cert-manager-kustomization.yaml
